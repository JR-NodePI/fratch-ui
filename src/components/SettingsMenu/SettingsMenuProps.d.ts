import { SettingsMenuPosition } from './SettingsMenuConstants';

export type SettingsMenuProps = {
  className?: string;
  items?: JSX.Element[];
  position?: (typeof SettingsMenuPosition)[keyof typeof SettingsMenuPosition];
};
