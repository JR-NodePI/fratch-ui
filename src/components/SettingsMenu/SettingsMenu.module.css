.settings_menu {
  z-index: 100;
  margin: 8px;
  padding: 8px;
  display: flex;
  position: absolute;
}

.settings_menu .opener {
  font-size: 16px;
  text-decoration: none;
  padding: 0;
  background-color: transparent;
  border: none;
  cursor: pointer;
  align-self: self-start;
  justify-content: center;
  align-items: center;
  display: flex;
  align-items: center;
  height: 36px;
  outline: none;
}
.opener_icon {
  color: var(--ft-color-text);
}

.settings_menu.open {
  background-color: var(--ft-color-white);
  border-radius: var(--ft-button-default-border-radius);
  box-shadow: var(--ft-box-shadow-2);
}
.settings_menu.open .menu {
  display: block;
}
.settings_menu.close .menu {
  display: none;
}

.settings_menu .menu {
  list-style: none;
  margin: 0;
  padding: 0;
  align-self: center;
}
.settings_menu .menu li {
  display: block;
  padding: 2px 8px;
}

/*** position ***/
.settings_menu.left {
  left: 0;
}
.settings_menu.left .opener {
  order: 1;
}
.settings_menu.left .menu {
  order: 2;
  text-align: left;
}

.settings_menu.right {
  right: 0;
}
.settings_menu.right .opener {
  order: 2;
}
.settings_menu.right .menu {
  order: 1;
  text-align: right;
}
