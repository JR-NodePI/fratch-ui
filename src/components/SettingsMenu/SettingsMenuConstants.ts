export const SettingsMenuPosition = {
  LEFT: 'left',
  RIGHT: 'right',
} as const;
