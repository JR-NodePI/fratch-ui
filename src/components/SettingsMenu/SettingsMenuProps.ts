export const SettingsMenuPosition = {
  LEFT: 'left',
  RIGHT: 'right',
} as const;

export type SettingsMenuProps = {
  className?: string;
  items?: JSX.Element[];
  position?: (typeof SettingsMenuPosition)[keyof typeof SettingsMenuPosition];
};
