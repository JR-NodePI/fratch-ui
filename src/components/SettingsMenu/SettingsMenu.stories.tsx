import { createPortal } from 'react-dom';

import type { Meta, StoryObj } from '@storybook/react';

import { Button, InputCheck } from '..';
import SettingsMenu from './SettingsMenu';
import { SettingsMenuPosition } from './SettingsMenuProps';

const inlineStyles: React.CSSProperties = {
  position: 'fixed',
  zIndex: 999999,
  top: 0,
  width: '100%',
  height: '64px',
  backgroundColor: '#e6e6e6',
};

function StoryRenderer(args: StoryObj['args']) {
  return createPortal(
    <div style={inlineStyles}>
      <SettingsMenu {...args} />
    </div>,
    document.body
  );
}

const meta = {
  title: 'Example/SettingsMenu',
  component: StoryRenderer,
  tags: ['autodocs'],
  argTypes: {
    position: {
      options: Object.values(SettingsMenuPosition),
      control: { type: 'radio' },
    },
  },
} satisfies Meta<typeof SettingsMenu>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Docs: Story = {
  args: {
    position: 'left',
    items: [
      <a href="javascript:void(0)">Setting menu item 1</a>,
      <a href="javascript:void(0)">Setting menu item 2</a>,
      <Button>Setting menu item 3</Button>,
      <Button size="small" type="primary">
        Setting menu item 4
      </Button>,
      <Button size="small" type="secondary">
        Setting menu item 5
      </Button>,
      <Button size="small" type="tertiary">
        Setting menu item 6
      </Button>,
      <InputCheck label="Setting menu item 7" position="right" />,
    ],
  },
};
