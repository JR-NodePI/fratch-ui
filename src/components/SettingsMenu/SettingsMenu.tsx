import { c } from '../../helpers/classNameHelpers';
import { useEffect, useState } from 'react';
import { v4 as uuid } from 'uuid';
import { SettingsMenuPosition, SettingsMenuProps } from './SettingsMenuProps';
import { isAscendantEvenTargetByID } from '../../helpers/htmlSelectorsHelpers';

import styles from './SettingsMenu.module.css';

export default function SettingsMenu({
  className,
  items = [],
  position = SettingsMenuPosition.LEFT,
}: SettingsMenuProps): JSX.Element {
  const [openerWrapperId] = useState(uuid());
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    const handleClick = (event: MouseEvent): void => {
      if (!isAscendantEvenTargetByID(event, openerWrapperId)) {
        setVisible(false);
      }
    };
    window.document.body.addEventListener('click', handleClick, true);
    return () => {
      window.document.body.removeEventListener('click', handleClick);
    };
  }, [openerWrapperId]);

  const handleOpenerClick = (): void => {
    setVisible(!visible);
  };

  return (
    <div
      className={c(
        className,
        styles.settings_menu,
        visible ? styles.open : styles.close,
        styles[position]
      )}
      id={openerWrapperId}
    >
      <ul className={c(styles.menu)}>
        {items.map((item, index) => (
          <li key={index}>{item}</li>
        ))}
      </ul>
      <button onClick={handleOpenerClick} className={c(styles.opener)}>
        â‹®
      </button>
    </div>
  );
}
