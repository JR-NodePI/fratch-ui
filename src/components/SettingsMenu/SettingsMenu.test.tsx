import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import SettingsMenu from './SettingsMenu';
import { SettingsMenuProps } from './SettingsMenuProps';

describe('SettingsMenu', () => {
  const settingsMenuId = 'crypto-mock-random-UUID-1234';
  const setup = (
    props: SettingsMenuProps = {}
  ): ReturnType<typeof render> & { settingsMenu: Element | null } => {
    const { container, ...data } = render(<SettingsMenu {...props} />);
    return {
      container,
      ...data,
      settingsMenu: container.querySelector(`#${settingsMenuId}`),
    };
  };

  beforeEach(() => {
    vi.spyOn(crypto, 'randomUUID').mockReturnValue(settingsMenuId);
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should render with items', () => {
    const items = [
      <span>item 1</span>,
      <span>item 2</span>,
      <span>item 3</span>,
    ];
    const { container } = setup({ items });
    expect(container).toMatchSnapshot();
  });

  it('should open and then close', async () => {
    const { settingsMenu } = setup();

    await userEvent.click(screen.getByRole('button'));

    expect(settingsMenu?.classList).toMatchInlineSnapshot(`
      DOMTokenList {
        "0": "_settings_menu_241598",
        "1": "_open_241598",
        "2": "_left_241598",
      }
    `);

    await userEvent.click(window.document.body);

    expect(settingsMenu?.classList).toMatchInlineSnapshot(`
      DOMTokenList {
        "0": "_settings_menu_241598",
        "1": "_close_241598",
        "2": "_left_241598",
      }
    `);
  });

  it('should render with className', () => {
    const { settingsMenu } = setup({ className: 'mock_test_class_name' });

    expect(settingsMenu?.classList).toMatchInlineSnapshot(`
      DOMTokenList {
        "0": "mock_test_class_name",
        "1": "_settings_menu_241598",
        "2": "_close_241598",
        "3": "_left_241598",
      }
    `);
  });

  it('should render with right position', () => {
    const { settingsMenu } = setup({ position: 'right' });

    expect(settingsMenu?.classList).toMatchInlineSnapshot(`
      DOMTokenList {
        "0": "_settings_menu_241598",
        "1": "_close_241598",
        "2": "_right_241598",
      }
    `);
  });
});
