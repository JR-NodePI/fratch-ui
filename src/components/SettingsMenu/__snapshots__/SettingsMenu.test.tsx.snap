// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsMenu > should render properly 1`] = `
<div>
  <div
    class="_settings_menu_241598 _close_241598 _left_241598"
    id="crypto-mock-random-UUID-1234"
  >
    <ul
      class="_menu_241598"
    />
    <button
      class="_opener_241598"
    >
      <svg
        aria-hidden="true"
        class="_icon_2be8fc _primary_2be8fc _opener_icon_241598"
        fill="currentColor"
        viewBox="0 0 4 15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.5 1.5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Zm0 6.041a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Zm0 5.959a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`SettingsMenu > should render with items 1`] = `
<div>
  <div
    class="_settings_menu_241598 _close_241598 _left_241598"
    id="crypto-mock-random-UUID-1234"
  >
    <ul
      class="_menu_241598"
    >
      <li>
        <span>
          item 1
        </span>
      </li>
      <li>
        <span>
          item 2
        </span>
      </li>
      <li>
        <span>
          item 3
        </span>
      </li>
    </ul>
    <button
      class="_opener_241598"
    >
      <svg
        aria-hidden="true"
        class="_icon_2be8fc _primary_2be8fc _opener_icon_241598"
        fill="currentColor"
        viewBox="0 0 4 15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.5 1.5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Zm0 6.041a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Zm0 5.959a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
